### ISSUE TITLE
To dream the impossible dream...

### ISSUE THREAD

**Issue by phenaproxima** [grom358/pharborist#61]
So this would be something for way down the line, but Pharborist's filtering, while effective, is currently a bit cumbersome. To do complex filtering, you MUST write your own callback.

What would be amazing is for Pharborist to have its own query language, like CSS selectors work in jQuery. Obviously CSS has nothing to do with PHP, so this would be an entirely new syntax, but it should be simple and effective for filtering. I'm opening this issue so I have a place to dump ideas about this, even if we never get around to implementing them. This feature, IMO, would make Pharborist the best PHP parser on the planet, hands-down. (It's already the easiest to use.)

To start, I think filtering by node type using Pharborist query language should look something like this:

``` php
$collection->filter('is ClassNode')
```

Filtering by multiple class types:

``` php
$collection->filter('is ClassNode or FunctionDeclarationNode or StringNode')
```

Essentially, "is" is a direct mapping to instanceof or is_a(). I'm guessing that class name resolution would be the hurdle here; either that, or one other option is to have this "is" query dispatch the class names to a variadic filter function which knows to return TRUE if any class names match.

Further ideas...filtering by class definition or function name:

``` php
$collection->children('class MyCryptoClass')
$collection->children('function foobar');
```

...which are shortcuts of:

``` php
$collection->children('is ClassNode named MyCryptoClass')
$collection->children('is FunctionDeclarationNode named foobar');
```

Or function/method call:

``` php
$collection->find('call to my_wacky_function')
$collection->find('call to MyClass::makeFunStatic')
$collection->find('call to $obj->bitchinMethod')
```

...which are shortcuts of:

``` php
$collection->find('is FunctionCallNode and calls my_wacky_function')
$collection->find('is ClassMethodCallNode and calls MyClass::makeFunStatic')
$collection->find('is ObjectMethodCallNode and calls $obj->bitchinMethod')
```

Here are some filtering scenarios that have come up in DMU (in pseudocode).

``` php
$function->hasParameter('Drupal\\node\\NodeInterface $node') // type hint optional, returns boolean

// In query language (making this up as I go along)
$tree->find('is FunctionDeclarationNode named diff_view_node_diffs and has parameter $node of type Drupal\\node\\NodeInterface')

$function->reassigns('$node') // TRUE if the $node variable is fully reassigned anywhere in the function
$function->find('is AssignmentNode and assigns $node') // Query language equivalent

$assignment->getsPropertyOf('$node') // TRUE if the assignment looks like $var = $node->property
$function->find('is AssignmentNode and gets property of $node'); // Query language equivalent

$assignment->setsPropertyOf('$node') // TRUE if the assignment looks like $node->property = $var
$function->find('is AssignmentNode and sets property of $node'); // Query language equivalent

$function->find('is AssignmentNode and gets property nid of $node');
$function->find('is AssignmentNode and sets property uid of $node');
```

In this model, the word "and" becomes a chaining operator between filtering steps. So a string like "is AssignmentNode and gets property of $node" is split by "and", and becomes two distinct operations: "is AssignmentNode", then "gets property of $node". Filter once, then filter again. Like faceted search.

To find subclasses or classes that implement an interface:

``` php
$tree->find('is ClassNode and extends GrandDaddyClass');
$tree->find('is ClassNode and implements \\ArrayAccess or \\Countable');
```

To find abstract/final classes:

``` php
$collection->find('is ClassNode and abstract')
$collection->find('is ClassNode and final')
```

To find classes with a specific method/property:

``` php
$collection->find('is ClassNode and has method magicalMysteryTour')
$collection->find('is ClassNode and has property ringo')
$collection->find('is ClassNode and has protected method clearAll')
// "contains" is a synonym for "has"
$collection->find('is ClassNode and contains private property $cache')
```

One possible design here is that the filtering system itself has a tree-like structure. You start with the top-level filters, like "is". The results of the "is" filter reveal other filters, which are exposed by various node types. For example, AssignmentNode exposes filters for "assigns", "gets property of" and "sets property of". ClassNode exposes "named",  "extends", "implements", etc. filters; FunctionDeclarationNode exposes "named", etc.

Multiple filter chains could be separated by commas, exactly like in CSS.

``` php
$tree->find('class PharboristBadassery, function use_pharborist_or_be_square')
```

"Or" groupings could be done with parens, as in SQL:

```
(is ClassNode implements NodeHandlerInterface) or (is FunctionDeclarationNode with parameter $node of type Drupal\node\NodeInterface)
```

Possible syntax to filter on function calls that look a certain way:

``` php
// Any call to variable_get()
$collection->filter('call to variable_get') 

// Calls to variable_get() with at least one argument
$collection->filter('call to variable_get with arguments')

// Calls to variable_get that passes a StringNode and ScalarNode
$collection->filter('call to variable_get with arguments (StringNode, ScalarNode)')

// Calls to variable_get that has a StringNode as the first or second argument,
// as long as there are two arguments.
$collection->filter('call to variable_get with arguments (StringNode, *) or (*, StringNode)')

// Calls to call_user_func with at least two arguments of any type
$collection->filter('call to call_user_func with arguments(*, *, ...)')
```

An implementation idea: Node can get a filter() or naturalFilter() method, which it is supposed to use to perform a test on itself. Different node types can expose different filters, but all nodes have an 'is' filter.

``` php
// Given this query: "is FunctionCallNode"
// $filter is the name of the filter, e.g. 'is'
// $arguments is a string, with everything after 'is' until the next filter name. In this case, $arguments is "FunctionCallNode"
public function naturalFilter($filter, $arguments) {
  if ($filter == 'is') {
    return $this instanceof $arguments;
  }
}
```

Proposed syntax to negate a filter:

``` php
$collection->find('is ClassNode and not abstract')
$collection->find('is ClassNode and ~has property ringo') // ~ stands in for "not"
$tree->find('is FunctionDeclarationNode not named diff_view_node_diffs and has parameter $node of type Drupal\\node\\NodeInterface')
```

To keep things as English-like as possible, "with" can be an alias for "and has":

``` php
$tree->find('is FunctionDeclarationNode with parameter $node of type Drupal\\node\\NodeInterface and not named diff_view_node_diffs')
// 'without' could be an alias for 'and not has'
```

Along those lines, "but" could also be an alias for "and":

``` php
$tree->find('is FunctionDeclarationNode not named diff_view_node_diffs but has parameter $node of type Drupal\\node\\NodeInterface')
$tree->find('is FunctionDeclarationNode with parameter $node of type \\Drupal\\node\\NodeInterface but not named diff_view_node_diffs')
```

Regex support:

``` php
$tree->find('is FunctionDeclarationNode named like ^(diff|pants)_')
```

Tree position predicates:

``` php
// "child of" is a synonym for "with parent", a.k.a. "and has parent"
$tree->find('is CallNode and child of ReturnStatementNode')

// "descends from", "descending from" and "in" are synonyms for "descendant of"
$tree->find('is AddNode and descendant of ArrayNode')

// "parent of" is a synonym for "with child", a.k.a. "and has child"
$tree->find('is ArrayNode and parent of StringNode')

// "ancestor of" is a synonym for "with", a.k.a. "and has"
$tree->find('is ArrayNode and ancestor of SubtractNode')
```

Basically any methods that currently filter on a callback should be able to filter on a natural language query too. I predict that the more English-like it can be, the more users will freaking LOVE it. By no means is this very important right now (maybe not at all); it's a pie-in-the-sky dream.

Any thoughts on this?
